app/presenters/account_relationships_presenter.rb

```diff
@@ -5,11 +5,67 @@ class AccountRelationshipsPresenter
               :muting, :requested, :domain_blocking
 
   def initialize(account_ids, current_account_id, **options)
-    @following       = Account.following_map(account_ids, current_account_id).merge(options[:following_map] || {})
-    @followed_by     = Account.followed_by_map(account_ids, current_account_id).merge(options[:followed_by_map] || {})
-    @blocking        = Account.blocking_map(account_ids, current_account_id).merge(options[:blocking_map] || {})
-    @muting          = Account.muting_map(account_ids, current_account_id).merge(options[:muting_map] || {})
-    @requested       = Account.requested_map(account_ids, current_account_id).merge(options[:requested_map] || {})
-    @domain_blocking = Account.domain_blocking_map(account_ids, current_account_id).merge(options[:domain_blocking_map] || {})
+    @account_ids        = account_ids.map { |a| a.is_a?(Account) ? a.id : a }
+    @current_account_id = current_account_id
+
+    @following       = cached[:following].merge(Account.following_map(@uncached_account_ids, @current_account_id))
+    @followed_by     = cached[:followed_by].merge(Account.followed_by_map(@uncached_account_ids, @current_account_id))
+    @blocking        = cached[:blocking].merge(Account.blocking_map(@uncached_account_ids, @current_account_id))
+    @muting          = cached[:muting].merge(Account.muting_map(@uncached_account_ids, @current_account_id))
+    @requested       = cached[:requested].merge(Account.requested_map(@uncached_account_ids, @current_account_id))
+    @domain_blocking = cached[:domain_blocking].merge(Account.domain_blocking_map(@uncached_account_ids, @current_account_id))
+
+    cache_uncached!
+
+    @following.merge!(options[:following_map] || {})
+    @followed_by.merge!(options[:followed_by_map] || {})
+    @blocking.merge!(options[:blocking_map] || {})
+    @muting.merge!(options[:muting_map] || {})
+    @requested.merge!(options[:requested_map] || {})
+    @domain_blocking.merge!(options[:domain_blocking_map] || {})
+  end
+
+  private
+
+  def cached
+    return @cached if defined?(@cached)
+
+    @cached = {
+      following: {},
+      followed_by: {},
+      blocking: {},
+      muting: {},
+      requested: {},
+      domain_blocking: {},
+    }
+
+    @uncached_account_ids = []
+
+    @account_ids.each do |account_id|
+      maps_for_account = Rails.cache.read("relationship:#{@current_account_id}:#{account_id}")
```

Comment:
`read_multi` may be used, may be not.

Reply:
I think with a limit on maximum requested relationships below 1000 there's almost no performance difference between read and read_multi, it would be a bit more bothersome to match back the values to the IDs..